src/envoy/http/error_translator/filter.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/envoy/http/error_translator/filter.h"
+
+#include <string>
+
+#include "common/http/headers.h"
+#include "external/envoy/include/envoy/stream_info/_virtual_includes/stream_info_interface/envoy/stream_info/stream_info.h"
+#include "src/envoy/utils/filter_state_utils.h"
+
+namespace espv2 {
+namespace envoy {
+namespace http_filters {
+namespace error_translator {
+
+using ::Envoy::Http::FilterHeadersStatus;
+
+Filter::Filter(FilterConfigSharedPtr config) : config_(config) {}
+
+// TODO(nareddyt): encodeHeaders or encodeData?
+// Probably need both for `handleEnvoyFilterError`
+FilterHeadersStatus Filter::encodeHeaders(Envoy::Http::ResponseHeaderMap&,
+                                          bool) {
+  if (utils::hasErrorFilterState(
+          *encoder_callbacks_->streamInfo().filterState())) {
+    // Enhanced handling of espv2 filter errors.
+    handleEspv2FilterError();
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (encoder_callbacks_->streamInfo().responseCodeDetails() !=
+      Envoy::StreamInfo::ResponseCodeDetails::get().ViaUpstream) {
+    // Basic handling of non-espv2 upstream envoy filter errors.
+    handleEnvoyFilterError();
+    return FilterHeadersStatus::Continue;
+  }
+
+  // Normal reply, not an error.
+  return FilterHeadersStatus::Continue;
+}
+
+void Filter::handleEspv2FilterError() {
+  google::rpc::Status error = utils::getErrorFilterState(
+      *encoder_callbacks_->streamInfo().filterState());
+
+  if (config_->shouldScrubDebugDetails()) {
+    // This is a copy, will not affect the filter state.
+    error.clear_details();
+  }
+
+  // TODO(nareddyt): respond with JSON.
+}
+
+void Filter::handleEnvoyFilterError() {
+  /*
+   * TODO(nareddyt)
+   *
+   * - Read the envoy filter error text from:
```

Comment:
For handleEnvoyError, we need a reliable way to identify the data from"sendLocalRelay",  and
if content-type is "plain/text" since the default "sendLocalReply"  is using this type, then,

* read the body, and wrap it to JSON as
{ 
  code: http_status
  message:  the body
}

we could not use grpc-status and grpc-message since it is on the trailers.  It is not easy to modify the body data in encodeTrailer() call

Reply:
We can follow this envoy code:

https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L732

Reply:
Since the trailers-only response uses `encodeHeaders`, I decided to also add logic for this. I think it's important, otherwise SC Report will be difficult. With the new logic here, we can just do an SC report using all the details in the filter state.

Let me know if that makes it too complex, we can remove it.